### ISSUE TITLE
Magento cannot run with custom Composer vendor directories

### ISSUE THREAD

**Issue by vbuck** [magento/magento2#17753]
When using the `COMPOSER_VENDOR_DIR` setting to specify a vendor path outside of the Magento installation root, Magento's autoloader registration fails to generate the correct path.

### Preconditions

1. Any host with access to two separate parent directories.
2. Any recent version of composer installed

### Steps to reproduce

1. Create a temporary directory for vendor data: `mkdir /tmp/composer-vendor-dir`
2. Create a project directory: `mkdir ~/magento-project`
3. Set the Composer vendor directory: `export COMPOSER_VENDOR_DIR="/tmp/composer-vendor-dir"`
4. Create a Magento project: `composer create-project --repository-url=https://repo.magento.com/ magento/project-community-edition ~/magento-project`
5. Observe successful Composer installation
6. Attempt to access Magento CLI: `php bin/magento`

### Expected result

1. Using `php bin/magento` should produce a command list

### Actual result

1. Error message given:
```
Autoload error: Vendor autoload is not found. Please run 'composer install' under application root directory.
```

### Additional Information
File `app/autoload.php` incorrectly assumes a vendor path relative to the project:
```
define('BP', dirname(__DIR__));
...
$vendorAutoload = BP . "/{$vendorDir}/autoload.php";
```

Which in the example scenario given produces a path like:
* `/home/user/magento-project//tmp/composer-vendor-dir/autoload.php`

The validation of `$vendorAutoload` should be modified to resolve variations of the assembled path, in a way that it can accept an absolute path returned by `app/etc/vendor_path.php`

**Comment by magento-engcom-team** [comment]
Hi @vbuck. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me {$VERSION} instance` 

where `{$VERSION}` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (2.2-develop +). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@vbuck do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by ghost** [comment]
@vbuck, thank you for your report.
We've acknowledged the issue and added to our backlog.

**Comment by annismckenzie** [comment]
You can get around this if you at least know the relative path to the vendor dir outside of the project (e.g. you put `"vendor-dir": "../composer_vendor"` into your composer.json file) – I got that far but am now stuck at the problem that the template loading mechanism disallows fetching view files from a directory that's outside the Magento root. It is a security feature but the vendor directory should be exempted as it's known and might be located anywhere on the filesystem (that is, if this issue is fixed).

**Comment by vbuck** [comment]
Yes, @annismckenzie, that's right: `vendor-dir` is an alternative parameter; however, my main problem is that the bootstrapper itself does not have any regard for absolute paths. It always assumes relative.

You are also right about the follow-up problem, in that views cannot be fetched outside of the Magento root. I hadn't considered `vendor` as an exemption from that feature but it makes sense to me, in that Magento should support all variations of Composer configurations.

**Comment by magento-engcom-team** [comment]
Hi @Rus0. Thank you for working on this issue. 
Looks like this issue is already verified and confirmed. But if your want to validate it one more time, please, go though the following instruction:
- [ ] 1. Add/Edit `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.

- [ ] 2. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 3. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 4. If the issue is not relevant or is not reproducible any more, feel free to close it.

**Comment by Rus0** [comment]
MM18MX

**Comment by magento-engcom-team** [comment]
Hi @vbuck. Thank you for your report. 
The issue has been fixed in magento/magento2#18849 by @Rus0 in 2.3-develop branch
Related commit(s):
 - [2054fab1731a63728a6b4e573cefbbf3e64d6c85](https://github.com/magento/magento2/commit/2054fab1731a63728a6b4e573cefbbf3e64d6c85)

The fix will be available with the upcoming 2.3.1 release.
